engine-src/Game/LambdaHack/Client/UI/SessionUI.hs

```diff
@@ -54,12 +54,7 @@ data SessionUI = SessionUI
                                     -- ^ parameters of the current run, if any
   , shistory       :: History       -- ^ history of messages
   , spointer       :: K.PointUI     -- ^ mouse pointer position
-  , slastAction    :: Maybe K.KM    -- ^ last pressed key
-  , smacroStack    :: MacroStack    -- ^ non-empty stack of macro buffers that
-                                    --   contains at least in-game macro buffer;
-                                    --   record keystrokes in the first Left
-                                    --   buffer
-  , slastPlay      :: KeyMacro      -- ^ state of key sequence playback
+  , sactionPending :: [ActionBuffer] -- ^ local action buffer
```

Comment:
I wonder, given that this list is always non-empty, could we separately store its head and tail here? Then also separately pass head and tail to all functions, avoiding error if the list empty?

Reply:
Thought about this. Certainly this would simplify the code, but I'm not sure if it's worth the effort. The thing is that for each action we would have to decide whether it was introduced by a macro, or was it the player. This gets even more complicated when player records a macro. Now we just don't care.

Reply:
Hmm, so would this simplify the code or complicate it, in the end? I can't see the problem without experimenting, but that's fine, feel free to merge and let me fail trying to refactor this later on. :)

Reply:
I meant it's a natural idea to separate these two and have one invariant less to take care of in the future. As for now, each time we handle an action, we're throwing an error if the stack is empty. That's rather poor design choice, since we wanted these functions to be pure, and it's totally avoidable if we pay the price of complicating the logic a bit more. I'm just not sure if it's worth of it.

Reply:
I see now: it simplifies the data invariant, but complicates the code. In that case, let's merge, to have one sensible point in the design space commited to the repo. Then we may try another one and compare.